### DISCUSSION TITLE
globset: Default for literal_separator

### DISCUSSION THREAD

**Discussion by mschoettle** [BurntSushi/ripgrep#2940] [Q&A] [answered]
I came across this through an application that uses the `globset` crate. By default `literal_separator` is `false` leading to `*` matching the path separator. My expectation was that `*` would not match path separators.

This also supported by the [man page for glob](https://man7.org/linux/man-pages/man7/glob.7.html):

> A '/' in a pathname cannot be matched by a '?' or '*' wildcard

What is the reason behind `literal_separator` being `false` by default?

**Comment by BurntSushi** [comment, marked as answer]
I followed [`glob`'s example](https://docs.rs/glob/latest/glob/struct.MatchOptions.html#method.new).

And it's also how Python's `fnmatch` module works:

```
>>> import fnmatch
>>> fnmatch.fnmatch('foo/bar/demo.txt', '*.txt')
True
```

I don't think there is any one commonly accepted default. And it depends on context. For a crate like `globset`, it's common to pass a full path to it. And in that context, it's _reasonable_ to expect that things like `*.txt` match `foo/bar.txt`. But in the context of something like `glob(7)`, it's more common to execute a glob in the context of a single directory where you don't expect `*` to traverse into sub-directories.

To be honest, globbing is a bit of a mess. And like regex, pretty much all implementations have their own little quirks and features.

**Reply by mschoettle** [reply]
Thanks for the detailed explanation!
